hammers/scripts/gpu_lease_stacking.py

```diff
@@ -8,80 +8,145 @@
 
 * ``info`` to just display leases or actuall delete them with ``delete``
 '''
-import sys
 import argparse
+import json
+import sys
 from pprint import pprint
 from datetime import datetime
 
 from hammers import MySqlArgs, osapi, query
 from hammers.slack import Slackbot
 from hammers.osrest.blazar import lease_delete
-
-LEASES_ALLOWED = 1
+from hammers.notifications import _email
 
 
-def find_stacked_leases(leases):
-    """Return list of only the leases stacked on each other."""
-    stacked = []
-
-    for i in range(len(leases)):
-
-        _, start_date, end_date = leases[i]
+LEASES_ALLOWED = 1
 
-        if i > 0:
-            last_end_date = leases[i - 1][2]
-        else:
-            last_end_date = datetime.min
 
-        if i < len(leases) - 1:
-            next_start_date = leases[i + 1][1]
-        else:
-            next_start_date = datetime.max
+class GPUUser:
+    """Class for user with advanced reservation on gpu nodes."""
 
-        stacked_previous = (start_date - last_end_date).days < 1
-        stacked_next = (next_start_date - end_date).days < 1
+    def __init__(self, user_id, name, email):
+        """Constructor."""
+        self.user_id = user_id
+        self.name = name
+        self.email = email
+        self.nodes = {}
+        self.leases_to_delete = []
 
-        if stacked_previous or stacked_next:
-            stacked.append(leases[i])
+    def add_lease(self, node_id, lease_id, start_date, end_date, **kwargs):
+        """Add lease to node list."""
+        if node_id not in self.nodes:
+            self.nodes[node_id] = []
 
-    return stacked
+        self.nodes[node_id].append((lease_id, start_date, end_date))
 
+    def delete_stacked_leases(self, auth):
+        """Delete leases in leases_to_delete."""
+        for lease_id, _, _ in self.leases_to_delete:
+            lease_delete(auth, lease_id)
 
-def reaper(db, auth, describe=False, quiet=False):
+    def sort_leases_by_date(self):
+        """Sort leases by date for each node."""
+        for node_id, leases in self.nodes.items():
+            self.nodes[node_id] = list(sorted(
+                set(leases), key=lambda x: x[1]))
+
+    def in_violation(self):
+        """Return boolean value for whether user has stacked gpu leases."""
+        return len(self.leases_to_delete) > 1
```

Comment:
Given that you only `extend` the leases to delete by a slice of `[LEASES_ALLOWED:]`, I think this should be checking that there are _any_ that have been added?

```suggestion
        return len(self.leases_to_delete) > 0
```

Reply:
You right